## 1 Импорт библиотек, создание подключения

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


## 2 Изучение данных

1) Рассмотрим, что из себя представляют таблицы БД, выявим возможные пропуски

2) Посмотрим на некоторые распределения

3) Поищем дубликаты

In [3]:
# запросы на загрузку таблицы из БД
query_1 = """
SELECT * FROM books;
"""
query_2 = """
SELECT * FROM authors;
"""
query_3 = """
SELECT * FROM publishers;
"""
query_4 = """
SELECT * FROM ratings;
"""
query_5 = """
SELECT * FROM reviews;
"""

In [4]:
books = pd.read_sql(query_1, engine)
authors = pd.read_sql(query_2, engine)
publishers = pd.read_sql(query_3, engine)
ratings = pd.read_sql(query_4, engine)
reviews = pd.read_sql(query_5, engine)

In [5]:
tables = ((books,'books'), (authors, 'authors'), (publishers, 'publishers'),
          (ratings,'ratings'), (reviews,'reviews'))

for df, i in tables:
    print(f'Таблица {i}')
    display(df.head())
    display(df.info())
    print(f'Дубликаты в таблице {i}:', df.duplicated().sum())
    print("=" * 40)
   

Таблица books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

Дубликаты в таблице books: 0
Таблица authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

Дубликаты в таблице authors: 0
Таблица publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

Дубликаты в таблице publishers: 0
Таблица ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

Дубликаты в таблице ratings: 0
Таблица reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

Дубликаты в таблице reviews: 0


Явные пропуски отсутсвуют, типы данных в целом соответствуют описанию, кроме атрибута publication_date, который object. При составлении запроса это нужно учесть и производить замену типа. 

Дубликаты отстутствуют.

Посмотрим на количественные параметры в таблицах

In [6]:
ratings['rating'].describe()

count    6456.000000
mean        3.928284
std         0.943303
min         1.000000
25%         3.000000
50%         4.000000
75%         5.000000
max         5.000000
Name: rating, dtype: float64

In [7]:
books['num_pages'].describe()

count    1000.00000
mean      389.11100
std       229.39014
min        14.00000
25%       249.00000
50%       352.00000
75%       453.00000
max      2690.00000
Name: num_pages, dtype: float64

Из расперделения количественных параметров видно, что оценки - оценки по 5-бальной шкале, количество страниц также в пределах разумного, аномалии, которые могут исказить результат, отсутствуют.

## Задача 1. Посчитайте, сколько книг вышло после 1 января 2000 года

Необходимо:

1) обратиться к таблице books
2) отфильтровать записи по publication_date больше 2000-01-01, преобразовав в date
3) посчитать количество по book_id

In [8]:
# запрос
query = """
SELECT COUNT(book_id)
FROM books
WHERE publication_date :: date > '2000-01-01';
"""
# загрузка
df = pd.read_sql(query, engine)
# вывод данных
df


count
0    819

In [9]:
# проверим
books['publication_date'] = books['publication_date'].astype('datetime64[D]')
display(books.query('publication_date > "2000-01-01"')['book_id'].count())
display(books.query('publication_date >= "2000-01-01"')['book_id'].count())

819

821

После 1 января 2000 (не включая) вышло 819 книг. Кто-то выпустил 2 книги в новый год=)

## Задача 2. Для каждой книги посчитайте количество обзоров и среднюю оценку

Необходимо:

1) объединить таблицы books, reviews, ratings по book_id
2) посчитать кол-во уникальных обзоров по review_id и найти среднюю оценку по rating (округим до 2 знака после запятой)
3) сгруппировать по book_id
4) выбрать поля book_id, title, кол-во обзоров, среднюю оценку

In [10]:
# запрос
query = """
SELECT b.book_id,
       b.title AS Title,
       COUNT(DISTINCT rev.review_id) AS Reviews_count,
       ROUND(AVG(rt.rating), 2) AS AVG_rating
FROM books AS b
LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
GROUP BY 1
ORDER BY 3 DESC;
"""
# загрузка
df = pd.read_sql(query, engine)
# вывод данных
df


book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     reviews_count  avg_rating  
0                7        3.66  
1                6        3.98  
2                6        4.21  
3                6        4.41  
4                6        4.08  
..             ...         ...  
995              0        3.67  
996              0        3.00  
997              0        5.00  
998              0        4.00  
999              0        4.00  

[1000 rows x 4 columns]

Сортировка в задаче не требовалась, но теперь мы знаем, что самые рецензируемые фильмы - Сумерки, Воды слонам, Стеклянный замок. Фанаты Роберта Патиссона зОхВаТиЛи чАтЕг

## Задача 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

1) Используем подзапрос в FROM,  в котором сформируем таблицу из объединенных таблиц books и publisher:
 - publisher_id
 - наименование издателя
 - количество книг данного издателя
 
2) В подзапросе FROM отфильтурем только те книги, в которых более 50 страниц, сгруппируем по наименованию/айди издателя, отсортируем по кол-ву книг, оставим 1-ю строку

Или можно использоввть врееную табицу и оконную функцию

In [11]:
# запрос
query = """
SELECT pb.publisher_id,
       pb.publisher,
       pb.cnt
FROM (SELECT pub.publisher_id AS publisher_id,
             pub.publisher AS publisher,   
             COUNT(b.book_id) AS cnt
      FROM books AS b
      LEFT JOIN publishers AS pub ON b.publisher_id = pub.publisher_id
      WHERE b.num_pages > 50
      GROUP BY pub.publisher_id
      ORDER BY COUNT(b.book_id) DESC
      LIMIT 1) AS pb
"""
# загрузка
df = pd.read_sql(query, engine)
# вывод данных
df

publisher_id      publisher  cnt
0           212  Penguin Books   42

или

In [12]:

query = """
WITH r AS
        (SELECT pub.publisher_id AS publisher_id,
                    pub.publisher AS publisher,   
                    COUNT(b.book_id) AS cnt,
                    ROW_NUMBER() OVER(ORDER BY COUNT(b.book_id) DESC) as rnk
            FROM books AS b
            LEFT JOIN publishers AS pub ON b.publisher_id = pub.publisher_id
            WHERE b.num_pages > 50
            GROUP BY pub.publisher_id)

SELECT r.publisher_id,
       r.publisher,
       r.cnt
FROM r
WHERE rnk=1;
"""

df = pd.read_sql(query, engine)

df

publisher_id      publisher  cnt
0           212  Penguin Books   42

Наибольшее число "толстых" книг выпустило издательство Penguin Books 

## Задача 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [13]:
# запрос
query = """
SELECT top.author
FROM (SELECT a.author as author,
             COUNT(rt.rating_id),
             ROUND(AVG(rt.rating), 2)
      FROM books AS b 
      LEFT JOIN authors AS a ON b.author_id=a.author_id
      LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
      GROUP BY 1
      HAVING COUNT(rt.rating_id) > 50
      ORDER BY 3 DESC
      LIMIT 1) AS top;

"""
# загрузка
df = pd.read_sql(query, engine)
# вывод данных
df

author
0  J.K. Rowling/Mary GrandPré

Автор с самым высоким рейтингом Джоан Роулинг, автор книг о Гарри Поттере. Мэри ГрандПре - художник-иллюстратор, создавшая знаменитые обложки серии. Среди данных есть еще одна книга за авторством Джоан Роулинг, но с гораздо меньшим количеством оценок.

При этом в поле автор указан Newt Scamander - это вообще то персонаж из книги Фантастические твари. Не совсем ясно, как он оказался среди авторов. Есть вопросики к базе. Ладно, если бы был указан псевдоним Роулинг - Роберт Гелтбрейт, например.

In [32]:
# запрос
query = """
SELECT top.author,
       top.cnt,
       top.avg_rating
FROM (SELECT a.author as author,
             COUNT(rt.rating_id) AS cnt,
             ROUND(AVG(rt.rating), 2) AS avg_rating
      FROM books AS b 
      LEFT JOIN authors AS a ON b.author_id=a.author_id
      LEFT JOIN ratings AS rt ON b.book_id=rt.book_id
      GROUP BY 1
      ORDER BY 3 DESC) top
WHERE top.author LIKE '%%Rowling%%';

"""
# загрузка
df = pd.read_sql(query, engine)
# вывод данных
df

author  cnt  avg_rating
0   J.K. Rowling/Mary GrandPré  312        4.29
1  Newt Scamander/J.K. Rowling    9        3.89

## Задача 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

Необходимо:
1) отобрать пользователей, которые поставили более 48 оценок для фильтарции в подзапросе
2) найти этих пользователей в таблице с рецензиями и посчитать их количесто (рецензий)
3) посчитать среднее


In [18]:
# запрос
query = """
SELECT ROUND(AVG(a.revs))
FROM (SELECT username,
             COUNT(review_id) AS revs
      FROM reviews 
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating_id) > 48)
      GROUP BY 1
      ORDER BY COUNT(review_id) DESC) AS a;

"""
# загрузка
df = pd.read_sql(query, engine)
# вывод данных
df

round
0   24.0